### PR TITLE
Fix computing metrics from periodically polled value with arbitrary polling period

### PR DESCRIPTION
**Pull request by Karakatiza666**
Is this a user-visible change (yes/no): yes

Fix: https://github.com/feldera/feldera/issues/491
> If I switch to a different tab and come back to the throughput graph some time later there is a fake throughput spike.

### CODE REVIEW

**Comment by Karakatiza666** [comment]
image
Added automatic Y axis scaling
Added axis labels
Normalizing throughput based on polling period

**Comment by mihaibudiu** [comment]
that's great!

**Comment by mihaibudiu** [comment]
maybe you can fix the conflicts too?

**Comment by Karakatiza666** [comment]
I'll update UI test snapshots later today

File: web-console/src/lib/components/streaming/management/PipelineTable.tsx [hunk 1]
```diff
@@ -312,10 +313,10 @@ const DetailPanelContent = (props: { row: Pipeline }) => {
           </Card>
         </Grid>
         <Grid item xs={5}>
-          <AnalyticsPipelineTput metrics={metrics.global} />
+          <AnalyticsPipelineTput metrics={metrics} keepMs={30000} />
```
Comment (mihaibudiu) [inline web-console/src/lib/components/streaming/management/PipelineTable.tsx:316]: can this value come from a global constant?
this way you don't risk to forget changing it in only one place

File: web-console/package.json [hunk 2]
```diff
@@ -23,7 +23,7 @@
     "format": "prettier --write \"{src,tests}/**/*.{js,jsx,ts,tsx}\"",
     "format:check": "prettier --check \"src/**/*.{js,jsx,ts,tsx}\"",
     "generate-openapi": "openapi --input ../openapi.json --output ./src/lib/services/manager",
-    "build-generate-openapi": "cargo run --bin pipeline-manager -- --dump-openapi && openapi --input ./openapi.json --output ./src/lib/services/manager && rm ./openapi.json",
+    "build-openapi": "cargo run --bin pipeline-manager -- --dump-openapi",
```
Comment (mihaibudiu) [inline web-console/package.json:26]: so the openapi.json file will be there or not?

**Review by snkas — COMMENTED** [review]
It's really nice to have the x-axis and y-axis, the graph visually looks good.

There are some UI interactions you should check for the graph to handle it correctly:
- Click away to another tab and come back
- Close- and open the pipeline information view
- Stopping the load-generation script shortly and then starting again

The x-axis is okay (it is clear that it is time due to the movement), possibly be better to have it in the "00:00" format (possibly "00:00:00 format"). The timestamp should also be based from pipeline start. The latter might be difficult. You could also consider fixed using just "0" on the right side, and then -10s, -20s, ..., -60s progressively to the left.

File: web-console/src/lib/components/streaming/management/AnalyticsPipelineTput.tsx [hunk 3]
```diff
@@ -3,33 +3,49 @@
...
+    throughput.at(-2) ?? throughput.at(-1) ?? tuple(0, 0),
+    throughput.at(-1) ?? tuple(0, 0)
+  )
   const series = [
     {
       data: throughput
     }
   ]
 
+  const valueMax = throughput.length ? Math.max(...throughput.map(v => v[1])) : 0
```
Comment (mihaibudiu) [inline web-console/src/lib/components/streaming/management/AnalyticsPipelineTput.tsx:37]: all these MUI tools and you have to do this by hand? there's no plotting tool which knows how to scale automatically?

File: web-console/src/lib/compositions/streaming/management/usePipelineMetrics.ts [hunk 4]
```diff
@@ -1,74 +1,62 @@
...
+      if (!_newData) {
+        setLastTimestamp(undefined)
+        return emptyData
+      }
+      const now = Date.now()
+      if (!lastTimestamp) {
+        setLastTimestamp(now)
+      }
+      const newData = toMetrics(_newData, { timeMs: lastTimestamp ? now - lastTimestamp : 0 })
+      const oldGlobal = oldData?.global ?? []
```
Comment (mihaibudiu) [inline web-console/src/lib/compositions/streaming/management/usePipelineMetrics.ts:45]: I don't think you should get the timestamps from the UI, they should come with the data.

File: demo/steady/demo.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        "description": "",
+        "code": program_sql
+    })
+    response.raise_for_status()
+    program_version = response.json()["version"]
+
+    # Compile program
+    print(f"Compiling program {program_name} (version: {program_version})...")
+    requests.post(f"{api_url}/v0/programs/{program_name}/compile", headers=headers, json={"version": program_version}).raise_for_status()
+    while True:
```
Comment (mihaibudiu) [inline demo/steady/demo.py:58]: this should be a utility function in some library

File: demo/steady/demo.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        print(f"Program status: {status}")
+        if status == "Success":
+            break
+        elif status != "Pending" and status != "CompilingRust" and status != "CompilingSql":
+            raise RuntimeError(f"Failed program compilation with status {status}")
+        time.sleep(5)
+
+    # Create pipeline
+    pipeline_name = "demo-steady-pipeline"
+    requests.put(f"{api_url}/v0/pipelines/{pipeline_name}", headers=headers, json={
```
Comment (mihaibudiu) [inline demo/steady/demo.py:69]: all these should be.
is there an issue for that?

File: demo/steady/demo.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    print(f"Inserting rows into table example...")
+    start_time_item_entry_generation = datetime.datetime.now().timestamp()
+    for entry_no in range(1, num_total + 1):
+        # Rate limiting
+        elapsed_s = datetime.datetime.now().timestamp() - start_time_item_entry_generation
+        target_generated = target_rate_per_second * elapsed_s
+        if float(entry_no - 1) > target_generated:
+            delta_too_much = float(entry_no - 1) - target_generated
+            sleep_time_s = delta_too_much / target_rate_per_second
+            print(f"Total elapsed time: {elapsed_s}")
```
Comment (mihaibudiu) [inline demo/steady/demo.py:96]: who reads this output

File: web-console/src/lib/components/streaming/management/PipelineMemoryGraph.tsx [hunk 8]
```diff
@@ -1,30 +1,43 @@
...
+export const PipelineMemoryGraph = (props: {
+  metrics: { global: (GlobalMetrics & { timeMs: number })[] }
+  keepMs: number
+}) => {
   const theme = useTheme()
 
-  const memUsed = props.metrics.map(m => m.rss_bytes ?? 0)
-  const smoothMemUsed = discreteDerivative(memUsed, (n1, n0) => n1 * 0.6 + 0.4 * n0)
+  const memUsed = props.metrics.global.map(m => tuple(m.timeMs, m.rss_bytes ?? 0))
+  const smoothMemUsed = ((n1, n0) => n1[1] * 0.5 + n0[1] * 0.5)(
```
[the remark this answers is not in the document]
Comment (Karakatiza666) [reply web-console/src/lib/components/streaming/management/PipelineMemoryGraph.tsx:19]: I believe it was introduced for cases with inconsistent throughput, as a value that changes significantly is more difficult to read

File: web-console/package.json (hunk 2 above)
Comment (Karakatiza666) [reply web-console/package.json:26]: I am leaving this command as a dev convenience only; the openapi.json is still supposed to be generated in the same PR where pipeline_manager API updated.

File: web-console/src/lib/components/streaming/management/AnalyticsPipelineTput.tsx (hunk 3 above)
Comment (Karakatiza666) [reply web-console/src/lib/components/streaming/management/AnalyticsPipelineTput.tsx:37]: I am using ApexCharts library for this; it does have auto-scaling, but it doesn't default to nice rounded min and max numbers, so I implemented this myself

File: web-console/src/lib/compositions/streaming/management/usePipelineMetrics.ts (hunk 4 above)
Comment (Karakatiza666) [reply web-console/src/lib/compositions/streaming/management/usePipelineMetrics.ts:45]: They don't yet =)

File: web-console/src/lib/components/streaming/management/PipelineMemoryGraph.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (snkas) [reply web-console/src/lib/components/streaming/management/PipelineMemoryGraph.tsx:19]: Sure, but memory usage is not over time, just a point in time at which it was that value

File: demo/steady/demo.py (hunk 6 above)
Comment (snkas) [reply demo/steady/demo.py:69]: There's not currently, it's a trade-off between requiring a dependency (library) and readability

Comment (snkas) [reply demo/steady/demo.py:69]: and understandability what is being done (i.e., just a HTTP request)

File: demo/steady/demo.py (hunk 5 above)
Comment (snkas) [reply demo/steady/demo.py:58]: (see other comment)

File: demo/steady/demo.py (hunk 6 above)
Comment (mihaibudiu) [reply demo/steady/demo.py:69]: and maintainability. If you ever change this, all the scripts will have to change.
it's clear we will need a library sooner or later, so we should build it

Comment (mihaibudiu) [reply demo/steady/demo.py:69]: but not for this pr

**Comment by Karakatiza666** [comment]
As there is no server-side metrics timestamps yet, I opted for relative, negative time labels:
image

File: demo/steady/demo.py (hunk 7 above)
Comment (Karakatiza666) [reply demo/steady/demo.py:96]: This script is designed to be run manually, so - the user in the CLI

**Comment by mihaibudiu** [comment]
Still not obvious whether these are seconds or minutes.
Why not just a number and a label of "seconds"?

**Comment by Karakatiza666** [comment]
I will update the displayed time on x-axis to an absolute time since pipeline start once this data is available with the metric in API

**Pull request merged**